### DISCUSSION TITLE
Regarding whether MooseFS 3.0.118 can be directly upgraded to 4.56.5?

### DISCUSSION THREAD

**Discussion by Leif160519** [moosefs/moosefs#599] [Q&A]
## Have you read through available documentation, open Github issues and Github Q&A Discussions?

Yes/No

#### Your moosefs version and its origin (moosefs.com, packaged by distro, built from source, ...).
|env|master ip|version|
|-|-|-|
|prod|192.168.31.110|3.0.118|
|test|192.168.31.120|4.56.5|
#### Operating system (distribution) and kernel version.
|env|master ip|os|kernel|
|-|-|-|-|
|prod|192.168.31.110|ubuntu20.04|5.4.0-193-generic|
|test|192.168.31.120|ubuntu 22.04|5.15.0-119-generic|
#### Hardware / network configuration, and underlying filesystems on master, chunkservers, and clients.
8c 16g 1000mbps 
#### How much data is tracked by moosefs master (order of magnitude)?
prod data size < 250G

![image](https://github.com/user-attachments/assets/508f871f-769d-4bb5-90cd-3eaf67446258)
![image](https://github.com/user-attachments/assets/ba1e6419-22ef-4858-8a42-5e7e57183290)

## Describe the problem you observed.

I have compiled and set up a new version of MooseFS cluster in the testing environment, which is 4.56.5. Personally, I also have a formal environment running MooseFS 3.0.118. I want to know if this 3.0.118 version can be directly upgraded to 4.56.5 by replacing the binary files and restarting the service. Additionally, I have a Kubernetes cluster that uses MooseFS 3.0.118 as its storage class. If it's possible to upgrade to version 4.56.5, will there be any impact on the Kubernetes cluster?

by the way,4.56.5 ui is amazing!!!

**Comment by Leif160519** [comment]
After compiling, I found that the executable file 'mfstools' was not generated, which means commands like 'mfsgetquota' and 'mfsgetsclass' cannot be used. The lower version of moosefs-client will also prompt 'not MFS object'. Do I need to wait for the official installation package to be released for support? Currently, https://repository.moosefs.com/moosefs-4/ is forbidden.
![image](https://github.com/user-attachments/assets/971ffd06-32d5-4bfb-9a32-29419267a6fb)

**Comment by chogata** [comment]
The new version does not have a `mfstools` executable anymore, the tools are now separate. Did you use the `linux_build.sh` script? Check out all the binaries that have been created in the `mfsclient` subdirectory, there is `mfsquota`, `mfssclass` and many more. If you run the install command `make install` as root, they all become available in your system. If not, you can use them directly via their path for testing purposes. The only difference is that after installing, aliases like `mfsgetquota` will become available, as long as you are just using binaries from the `msfclient` directory, you need to use the commands like `mfsquota get`, `mfsquota set` etc.

We are happy you like the new GUI :) And we are working on getting the package repository up and running.

Yes, you should be able to upgrade 3.0.118 directly to 4.56.5. As always, remember to back up your metadata, just in case. After you stop your old master, copy the `metadata.mfs` and all the` changelog.*.mfs` files to a backup location, then start the new master and it should load all your metadata without any problems.

Regarding Kubernetes: the driver for the MooseFS 3 version will not work with the client from MooseFS 4. You can upgrade your CGI, master, chunk servers and metaloggers, but you need to keep the old mount and tools for now to work with Kubernetes. You can install the new mount on another machine, just to have regular access to the filesystem. And as long as you connect a version 3 client, you definitely should not use the EC format for your chunks.
We are working on a driver that will be compatible with MooseFS 4, it shouldn't take long now.

**Comment by Leif160519** [comment]
> The new version does not have a `mfstools` executable anymore, the tools are now separate. Did you use the `linux_build.sh` script? Check out all the binaries that have been created in the `mfsclient` subdirectory, there is `mfsquota`, `mfssclass` and many more. If you run the install command `make install` as root, they all become available in your system. If not, you can use them directly via their path for testing purposes. The only difference is that after installing, aliases like `mfsgetquota` will become available, as long as you are just using binaries from the `msfclient` directory, you need to use the commands like `mfsquota get`, `mfsquota set` etc.
> 
> We are happy you like the new GUI :) And we are working on getting the package repository up and running.
> 
> Yes, you should be able to upgrade 3.0.118 directly to 4.56.5. As always, remember to back up your metadata, just in case. After you stop your old master, copy the `metadata.mfs` and all the` changelog.*.mfs` files to a backup location, then start the new master and it should load all your metadata without any problems.
> 
> Regarding Kubernetes: the driver for the MooseFS 3 version will not work with the client from MooseFS 4. You can upgrade your CGI, master, chunk servers and metaloggers, but you need to keep the old mount and tools for now to work with Kubernetes. You can install the new mount on another machine, just to have regular access to the filesystem. And as long as you connect a version 3 client, you definitely should not use the EC format for your chunks. We are working on a driver that will be compatible with MooseFS 4, it shouldn't take long now.

Thank you very much for your reply. Today, I found that the apt source of MooseFS-4 can be updated and installed, and I successfully upgraded it to version 4.56.6. However, I noticed that there is also a new version of MooseFS-CSI-Plugin. Could you please advise me which image I should use instead?
```
curl -X GET https://registry.moosefs.com/v2/moosefs-csi-plugin/tags/list | jq


{
  "name": "moosefs-csi-plugin",
  "tags": [
    "latest",
    "0.9.6-dev",
    "0.9.4-dev",
    "dev",
    "0.9.2-4.42.0-pro",
    "latest-dev",
    "0.9.4-3.0.117",
    "0.9.2-4.32.0-pro",
    "0.9.2-4.36.4-pro",
    "0.9.6-3.0.117",
    "0.9.2-dev",
    "0.9.6-4.49.0-pro",
    "0.9.4-3.0.117-pro",
    "0.9.6-3.0.117-pro",
    "0.9.1",
    "0.9.7-3.0.118-pro",
    "0.9.2-3.0.115",
    "0.9.2-4.38.4-pro",
    "0.9.5-dev",
    "0.9.6-3.0.118",
    "0.9.4-4.44.4-pro",
    "0.9.2-4.30.2",
    "0.9.2-3.0.116",
    "0.9.2-3.0.116-pro",
    "0.9.2-4.37.5-pro",
    "0.9.7-4.56.3-pro"
  ]
}
```
@chogata

**Comment by xandrus** [comment]
Hi @Leif160519

Can you try this driver with some fixes that I made recently?
https://github.com/xandrus/moosefs-csi/tree/dev

I would also like to add that the CSI image with the MooseFS 4.56.6 CE client is publicly available on ghrc:
https://github.com/users/xandrus/packages/container/package/moosefs-csi

If everything works flawlessly, I will try to push these changes to the main moosefs-csi repository.

**Comment by Leif160519** [comment]
> Hi @Leif160519
> 
> Can you try this driver with some fixes that I made recently? https://github.com/xandrus/moosefs-csi/tree/dev
> 
> I would also like to add that the CSI image with the MooseFS 4.56.6 CE client is publicly available on ghrc: https://github.com/users/xandrus/packages/container/package/moosefs-csi
> 
> If everything works flawlessly, I will try to push these changes to the main moosefs-csi repository.

When i use `ghcr.io/xandrus/moosefs-csi-plugin:dev`,it works!
```
[root@k3s-master-01:~ 16:45:29] # kubectl get sc | grep moosefs
moosefs-storage (default)   csi.moosefs.com         Delete          Immediate              true                   21m
[root@k3s-master-01:~ 16:45:36] # kubectl get pvc -n devops 
NAME          STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      VOLUMEATTRIBUTESCLASS   AGE
nginx-log     Bound    pvc-7e535f2a-14b8-4a64-bf51-6fd0de26e55f   1Gi        RWX            moosefs-storage   <unset>                 7m14s
nginx-vhost   Bound    pvc-5cf37af4-d2aa-4e24-aeec-673787923fee   1Gi        RWX            moosefs-storage   <unset>                 7m14s
```

```
tree
.
├── logs
│   ├── controller.log
│   ├── k3s-master-01.log
│   ├── k3s-node-01.log
│   └── k3s-node-02.log
└── volumes
    ├── pvc-5cf37af4-d2aa-4e24-aeec-673787923fee
    └── pvc-7e535f2a-14b8-4a64-bf51-6fd0de26e55f
        ├── access.log
        └── error.log
```

![image](https://github.com/user-attachments/assets/2ae85d4a-5df5-4c34-a7ff-b32d7400d630)

------------------------------------------------------------------------

When i use `ghcr.io/xandrus/moosefs-csi:dev`


![image](https://github.com/user-attachments/assets/345c2fb3-dcd0-4d10-b971-1a7b0091269e)

**Comment by chogata** [comment]
Just dropping a note here: MooseFS 4 packages are now available in the official repository, if anyone wants to use them rather than build from the sources.

**Comment by araujobsd** [comment]
Shall I use the same procedure to upgrade from 3.0.118 to 4.56.5 like described here ?https://moosefs.com/Content/Downloads/moosefs-upgrade.pdf

Is there any new process for upgrade? 

I currently have a production cluster with 48 hosts each with 2 NVME (chunkservers), 1 Master and 2 Metaloggers. This cluster is running KVM VMs.

Best,

**Reply by chogata** [reply]
The order of upgrade is still the same as in the linked document.
